rfcs/97-add-restrictions-in-storage-metadata.md

```diff
@@ -0,0 +1,148 @@
+---
+author: JinnyYi <github.com/JinnyYi>
+status: draft
+updated_at: 2021-06-09
+---
+
+# GSP-97: Add Restrictions In Storage Metadata
+
+## Background
+
+We use `Write` to handle single write operation. And the size is not unlimited.
+
+- S3 alike storage services have 5GB limit for a single PUT operation.
+- azblob limited to 5000 MiB for put blob (<https://docs.microsoft.com/en-us/rest/api/storageservices/put-blob>).
+
+If we upload a file with size out of limit in a single operation, we will get an error like `Request Entity Too Large` in azblob.
+
+Similar restriction also with `copy`,`append`, etc. We should figure out the error before sending request to reduce the consumption of network resources.
+
+These restrictions belong to storage. Currently, we have `StorageMeta` to carry storage metadata which is generated by `go generate cmd/definitions`:
+
+```go
+type StorageMeta struct {
+	location string
+	Name     string
+	WorkDir  string
+	
+	// bit used as a bitmap for object value, 0 means not set, 1 means set 
+	bit uint64
+	m   map[string]interface{}
+}
+```
+
+But the current `StorageMeta` cannot solve the above problem:
+
+- The fields are finite. For now, `m` is unused, and it's not convenient to hold the definite metadata.
+- Services have to call `metadata` repeatedly before use storage metadata internally.
+
+## Proposal
+
+So I propose to add restrictions in storage metadata.
+
+### Add global storage metadata
+
+Add maximum size restrictions as storage metadata into [info_storage_meta.toml]
+
+- `copy-size-maximum`: Maximum size for copy operation.
+- `fetch-size-maximum`: Maximum size for fetch operation.
+- `move-size-maximum`: Maximum size for move operation.
+- `write-size-maximum`: Maximum size for write operation.
+- `append-number-maximum`: Max append numbers in append operation.
+- `append-size-maximum`: Max append size in per append operation.
+- `append-total-size-maximum`: Max append total size in append operation.
+- `multipart-number-maximum`: Maximum part number in multipart operation.
+- `multipart-size-maximum`: Maximum part size defined by storage service.
+- `multipart-size-minimum`: Minimum part size defined by storage service.
+
+### Deprecate the Existing Storage Metadata in Object Metadata
+
+The following append and multipart related metadata in the current object metadata, actually belongs to storage metadata, should be deprecated in object metadata.
```

Comment:
Do I understand correctly?
- Before, we have some restrictions in object metadata for operations like `append` and `multipart`, but some restrictions are missing like those for `write`.
- In this proposal, we will move them from object metadata into storage metadata, and add some missing ones.

If so, why should we move them? Is this only a conceptual problem or does it bring some practical benefits?

Reply:
There are two kinds of service metadata here:

- Some of them are object-related state, like `append-offset` or `multipart-id`
- Some of them are storage level restrictions but misplaced at the object level, like `multipart-size-maximum`.

Storage level restrictions are shared across the whole service. The current design leads to following problems:

- Enforce users must get an Object to check the maximum, they can't check before creating a multipart object.
- Service should set those storage level restrictions every time creates an object, which makes it hard to maintain and does not meet our requirements about `zero-abstriction`.

Reply:
Restrictions for `append`， `multipart` are not for a particular object but for the service. We classify them as storage metadata, so we intend to move them from object metadata into storage metadata.